flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java

```diff
@@ -267,18 +267,31 @@ private void registerCallback() {
 
     @Override
     public void write(InputT element, Context context) throws IOException, InterruptedException {
+        while (mailboxExecutor.tryYield()) {}
```

Comment:
This seems weird 🤔 If I understand it correctly in means something along the lines of ... don't write results until we have anything else to do, so for example if user has a super frequent checkpointing (edge case), we can not make progress?

@pnowojski Can you please verify?

Reply:
Kind of @dmvk . One small caveat is that "yield" means "yield to downstream". So it only yields to mailbox actions enqueued by self OR downstream operators.

I have however a couple of suspicious about this:
1. We shouldn't be yielding on the hot path. Only if internal operator's buffer is full 
2. I don't understand why it's `tryYield()` in the first place? IMO It should have been `yield()` the loop below from the get go:
```
        while (bufferedRequestEntries.size() >= maxBufferedRequests) {
            mailboxExecutor.yield();
        }
```
Otherwise we have active/busy waiting loop wasting CPU resources.


A good example how I believe similar pattern is correctly implemented is `AsyncWaitOperator`.

I am kind of lost here and I would need to understand why it was changed to `tryYield` as part of https://issues.apache.org/jira/browse/FLINK-24234. I'm afraid that I won't be able to dig into this until next week.

Reply:
Thanks for the review and explanation. The current implementation is unnecessarily applying backpressure (FLINK-25792). I propose we take this feedback and remove the busy loop by migrating to `yield()` as stated. We will also review the rest of this code (`flush()` etc) with the above feedback in mind. We can then get this merged and raise a follow up Jira to re-review next week with @pnowojski. @dmvk how does that sound?

Reply:
Hi @pnowojski, thanks, just on the questions:
1. There are two types of scenarios where we enqueue to the mailbox (1) to handle fatal exceptions and (2) to add to the buffer any failed request entries. I believe, these should take priority over flushing new items?
2. I do agree the line is dubious, perhaps, this is more appropriate:
```
    while (inFlightRequestsCount > 0) {
        mailboxExecutor.yield();
    }
```
3. Perhaps I'm mistaken, but I don't believe we have a busy loop here. i.e. if `mailboxExecutor.tryYield()` returns true, there is some work to be elsewhere in the mailbox, then we perform that. Otherwise, it will return false and the loop will end. I can't see where CPU resources is being wasted?

Reply:
@dannycranmer It sounds fishy and doesn't align with the quality standards we require for Flink contributions. If you're concerned about the feature freeze, this can still be merged after that as the current implementation is broken.

Reply:
Just wanted to back up my feeling that I don't feel we have a busy loop on the [tryYield()](https://nightlies.apache.org/flink/flink-docs-master/api/java//org/apache/flink/streaming/runtime/tasks/mailbox/MailboxExecutorImpl.html#tryYield--):
set number of mailbox threads, call to tryyield will run the first one, and return true if it ran, until there are no more, when it will return false

Reply:
Maybe to rephrase that a bit, so it doesn't come out wrong: we shouldn't rush things that require more context from people that are busy with other efforts, just for sake of getting something merged before the feature freeze

Reply:
@dmvk ack, we will revisit this and get the PR into a state we are happy with and wait for additional reviews before merging. Thanks for the help

Reply:
@dmvk Would you mind letting us know what technical issues you feel there are here? I would be happy to address them and make fixes if necessary.

To answer your previous question, the semantic is don't buffer or write anything if there are any failed requests waiting to be requeued or fatal exceptions to fail the app with. If the user has super frequent checkpointing, the async threads will be taking care of writing to the destination, and the buffering will proceed as normal and not block here.

Reply:
Sorry for a bit late response.

> 1. There are two types of scenarios where we enqueue to the mailbox (1) to handle fatal exceptions and (2) to add to the buffer any failed request entries. I believe, these should take priority over flushing new items?

Apart of that, (2) has also a very important purpose to mark decrease the number of "in flight" requests, so that `AsyncSinkWriter` can induce back pressure to upstream operators.

> I do agree the line is dubious, perhaps, this is more appropriate: (the behaviour would remain identical)
```
while (inFlightRequestsCount > 0) {
  mailboxExecutor.yield(); 
}
```
I think that would be better, BUT that would block writes and backpressure upstream operators even if there is just a single element still "in flight" and the internal buffers are mostly empty. I'm not sure if that's the intention?

Can you elaborate what was wrong with the original condition
```
        while (bufferedRequestEntries.size() >= maxBufferedRequests) {
            mailboxExecutor.yield();
        }
```
?
I mean I would normally expected the following behaviour. Given some max buffer capacity, having two thresholds and the following conditions:
a) start flushing if `size >= threshold_1`
b) block new writes/flushes once if `size == max_capacity`
c) unblock if `size < threshold_2`

Where for example `AsyncWaitOperator` implicitly defines `threshold_1 = 1` and `threshold_2 = max_capacity`. But having those thresholds as some fractions of max capacity could improve performance by limiting the amount of small write requests.

With the `inFlightRequestsCount > 0`  condition to unblock, we are actually guaranteeing that the operator will be always wasting a bit of IO resources, as in this sequence of events:
1. external system finishing the last IO write
2. information about this reaching Flink, decreasing `inFlightRequestsCount` to `0`.
3. Flink sending another write request to the external system
4. external system starting it's own IO write.

there will be a bit of delay between 1. and 4.

> Perhaps I'm mistaken, but I don't believe we have a busy loop here. i.e. if mailboxExecutor.tryYield() returns true, there is some work to be elsewhere in the mailbox, then we perform that. Otherwise, it will return false and the loop will end. I can't see where CPU resources is being wasted?

Yes, this particular snippet in this PR  doesn't have busy waiting. The version in the current master has actually this exact problem. However as far as I understand it, `AsyncSinkWriter#flush` is doing the busy waiting both on master and in this PR?

On a side note, why do we need to have this logic spread among different methods? Can not we have one single check
```
while (someCondition()) {
  mailboxExecutor.yield();
}
```
?

Reply:
Thanks @pnowojski will start work on it next week.